modules/OFStateManager/module/src/gentable_handlers.c

```diff
@@ -289,14 +314,106 @@ ind_core_bsn_gentable_entry_add_handler(
         indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
                                            OF_BSN_GENTABLE_ERROR_TABLE_FULL,
                                            err_txt);
-    } else {
+    } else if (rv == INDIGO_ERROR_TIME_OUT) {
+        indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
+                                           OF_BSN_GENTABLE_ERROR_UNKNOWN,
+                                           err_txt);
+    } else if (rv < 0) {
         indigo_cxn_send_bsn_gentable_error(cxn_id, obj, table_id,
                                            OF_BSN_GENTABLE_ERROR_UNKNOWN,
                                            "Gentable add failed");
     }
+done:
+    return;
 }
 
-void
+indigo_error_t
+ind_core_bsn_gentable_entry_add_handler(
+    of_object_t *obj,
+    indigo_cxn_id_t cxn_id)
+{
+    uint16_t table_id;
+    indigo_core_gentable_t *gentable;
+    of_list_bsn_tlv_t key, value;
+    void *priv = NULL;
+    struct ind_core_gentable_entry *entry;
+    of_desc_str_t err_txt = "Gentable add failed";
+    indigo_error_t rv = INDIGO_ERROR_NONE;
+
+    of_bsn_gentable_entry_add_table_id_get(obj, &table_id);
+
+    gentable = find_gentable_by_id(table_id);
+    if (gentable == NULL) {
+        AIM_LOG_ERROR("Nonexistent gentable id %d in gentable_entry_add message", table_id);
+        indigo_cxn_send_error_reply(
+            cxn_id, obj,
+            OF_ERROR_TYPE_BAD_REQUEST,
+            OF_REQUEST_FAILED_BAD_TABLE_ID);
+        return rv;
+    }
+
+    /* bind key and value */
+    of_bsn_gentable_entry_add_key_bind(obj, &key);
+    of_bsn_gentable_entry_add_value_bind(obj, &value);
+
+    entry = find_entry_by_key(gentable, &key);
+
+    if (entry == NULL) {
+        if (gentable->ops->add4 != NULL) {
+            indigo_core_op_context_t *op_ctx = aim_zmalloc(sizeof(*op_ctx));
+            op_ctx->cxn_id = cxn_id;
+            op_ctx->obj = obj;
+            rv = gentable->ops->add4(op_ctx, gentable->priv, &key, &value, &priv, err_txt);
+            if (rv == INDIGO_ERROR_PENDING) {
+                /* entry hasn't been allocated yet, wait for resume.
+                 * block async op pending */
+                if (priv != NULL) {
+                    AIM_LOG_ERROR("gnetable %s return pending. Priv should be NULL", 
+                                  gentable->name);
+                }
+                indigo_cxn_block_async_op(cxn_id);
```

Comment:
since we're not copying the OF object, let's add a note that this implementation will only handle one pending operation at any point in time.
also, this "one-at-a-time" only should be called out in the same place where we're documenting limitations and design decisions for the current approach, so that we can revisit as necessary.